dist/performance.js

```diff
@@ -227,60 +239,141 @@ function calcOcclusion(rect){
     return Math.max(0, calcArea(clipped_rect)/calcArea(rect))
 }
 
+function splitSrcSet(srcset) {
+    // "img.jpg 100w, img2.jpg 300w"
+    return srcset.split(',').map(srcDesc => {
+        // "img.jpg 100w", " img2.jpg 300w"
+        const src = srcDesc.trim().split(/\s+/)[0];
+        return new URL(src, location.href).href;
+    });
+}
+
+function parseLinkHeader(link) {
+    if (!link) {
+        return {};
+    }
+
+    const srcPattern = /<([^>]+)>/;
+    const paramPattern = /([^=]+)=['"]?([^'"]+)['"]?/;
+    return Object.fromEntries(link.split(/[,\n]/).map(l => {
+        let [src, ...params] = l.split(';');
+
+        if (!srcPattern.test(src)) {
+            return [];
+        }
+        src = src.match(srcPattern)[1];
+        src = new URL(src, location.href).href;
+
+        params = params.map(p => {
+            p = p.trim().toLowerCase();
+            if (!paramPattern.test(p)) {
+                return;
+            }
+            const [_, key, value] = p.match(paramPattern);
+            return {key, value};
+        });
+        return [src, params];
+    }));
+}
+
+function getLcpPreloadInfo(rawDoc, lcpUrl) {
+    const lcpPreload = {};
+
+    const preloadTag = Array.from(rawDoc.querySelectorAll('link')).find(link => {
+        if (link.rel != 'preload') {
+            return false;
+        }
+
+        let src = link.href;
+        if (link.hasAttribute('imagesrcset')) {
+            src = splitSrcSet(link.imagesrcset).find(src => src == lcpUrl);
+        }
+
+        return src == lcpUrl;
+    });
+
+    if (preloadTag) {
+        lcpPreload.tag = summarizeElement(preloadTag);
+    }
+
+    const linkHeaderString = response_bodies[0].response_headers.link;
+    if (!linkHeaderString) {
+        return lcpPreload;
+    }
+
+    const directives = parseLinkHeader(linkHeaderString);
+    if (!(lcpUrl in directives)) {
+        return lcpPreload;
+    }
+
+    const isLcpPreloadedInLinkHeader = directives[lcpUrl].some(param => {
+        return param.key == 'rel' && param.value == 'preload';
+    });
+
+    if (isLcpPreloadedInLinkHeader) {
+        lcpPreload.header_directives = Object.fromEntries(directives[lcpUrl].map(d => {
+            return [d.key, d.value];
+        }));
+    }
+
+    return lcpPreload;
+}
+
+function getRawLcpElement(rawDoc, lcpUrl) {
+    const rawLcpElement = Array.from(rawDoc.querySelectorAll('picture source, img')).find(i => {
+        let src = i.src;
+        if (i.hasAttribute('srcset')) {
+            src = splitSrcSet(i.srcset).find(src => src == lcpUrl);
+        }
+
+        return src == lcpUrl;
```

Comment:
If you are not adding text elements, then this might be null/none/undefined.

```suggestion
        return src && src == lcpUrl;
```

Reply:
This function is only called when `lcpUrl != ''` (see `isLcpExternalResource`)

Reply:
Ah missed that. Resolving.